---
id: PERF-ANIMATION-NO-SETSTATE
title: "–ê–Ω–∏–º–∞—Ü–∏–∏: –∏—Å–ø–æ–ª—å–∑—É–π—Ç–µ AnimatedBuilder –≤–º–µ—Å—Ç–æ setState"
status: stable
severity: warning
category: dart/performance
tags: [performance, animation, setState, rebuild, AnimatedBuilder]
version: 2
owners: ["@team-performance"]
bad_snippet: bad-animation-setstate-001.dart
good_snippet: good-animation-builder-001.dart
references:
  - https://docs.flutter.dev/perf/best-practices
  - https://api.flutter.dev/flutter/widgets/AnimatedBuilder-class.html
  - https://api.flutter.dev/flutter/rendering/CustomPainter/CustomPainter.html
  - https://api.flutter.dev/flutter/rendering/FlowDelegate/FlowDelegate.html
  - https://api.flutter.dev/flutter/rendering/MultiChildLayoutDelegate/MultiChildLayoutDelegate.html
linter_rule:
  coverage: ai
ai_hint: >
  Flag AnimationController.addListener() with setState() calls inside.
  This causes full widget rebuild on every animation frame (60 FPS).
  Suggest using AnimatedBuilder, specialized transition widgets (RotationTransition, FadeTransition, etc.),
  CustomPaint with repaint parameter, FlowDelegate with repaint, or MultiChildLayoutDelegate with relayout.
  Explain that repaint/relayout skip Build and Element phases, going directly to Paint/Layout.
  Suggest Listenable.merge() when multiple independent sources need to be tracked.
  Do not flag: setState in user input handlers, one-time animations, or non-animation contexts.
---

## –ü–æ—è—Å–Ω–µ–Ω–∏–µ

–í—ã–∑–æ–≤ `setState()` –≤ `addListener()` –∞–Ω–∏–º–∞—Ü–∏–æ–Ω–Ω–æ–≥–æ –∫–æ–Ω—Ç—Ä–æ–ª–ª–µ—Ä–∞ –ø—Ä–∏–≤–æ–¥–∏—Ç –∫ **–ø–æ–ª–Ω–æ–º—É rebuild** –≤–∏–¥–∂–µ—Ç–∞ **–Ω–∞ –∫–∞–∂–¥–æ–º –∫–∞–¥—Ä–µ –∞–Ω–∏–º–∞—Ü–∏–∏** (60 —Ä–∞–∑ –≤ —Å–µ–∫—É–Ω–¥—É).

–≠—Ç–æ —Å–æ–∑–¥–∞–µ—Ç –æ–≥—Ä–æ–º–Ω—É—é –Ω–∞–≥—Ä—É–∑–∫—É –Ω–∞ CPU, –ª–∏—à–Ω–µ–µ –¥–∞–≤–ª–µ–Ω–∏–µ –Ω–∞ GC –∏ –º–æ–∂–µ—Ç –ø—Ä–∏–≤–µ—Å—Ç–∏ –∫ –ø—Ä–æ–ø—É—Å–∫–∞–º –∫–∞–¥—Ä–æ–≤ (jank).

**–†–µ—à–µ–Ω–∏–µ:** –ò—Å–ø–æ–ª—å–∑—É–π—Ç–µ `AnimatedBuilder`, —Å–ø–µ—Ü–∏–∞–ª–∏–∑–∏—Ä–æ–≤–∞–Ω–Ω—ã–µ –≤–∏–¥–∂–µ—Ç—ã –∏–ª–∏ `CustomPaint` —Å –ø–∞—Ä–∞–º–µ—Ç—Ä–æ–º `repaint`.

## –ü–æ—á–µ–º—É —ç—Ç–æ –∫—Ä–∏—Ç–∏—á–Ω–æ

- `setState()` –≤—ã–∑—ã–≤–∞–µ—Ç **–ø–æ–ª–Ω—ã–π rebuild** –≤—Å–µ–≥–æ –≤–∏–¥–∂–µ—Ç–∞ –∏ –µ–≥–æ –¥–µ—Ç–µ–π
- –ü—Ä–æ–∏—Å—Ö–æ–¥–∏—Ç **60 —Ä–∞–∑ –≤ —Å–µ–∫—É–Ω–¥—É** (–∫–∞–∂–¥—ã–π –∫–∞–¥—Ä)
- –ü–µ—Ä–µ—Å–æ–∑–¥–∞—é—Ç—Å—è –≤—Å–µ –¥–æ—á–µ—Ä–Ω–∏–µ –≤–∏–¥–∂–µ—Ç—ã
- –û–≥—Ä–æ–º–Ω–∞—è –Ω–∞–≥—Ä—É–∑–∫–∞ –Ω–∞ CPU
- –õ–∏—à–Ω–µ–µ –¥–∞–≤–ª–µ–Ω–∏–µ –Ω–∞ GC ‚Üí –ø—Ä–æ–ø—É—Å–∫–∏ –∫–∞–¥—Ä–æ–≤

## ‚ùå –ü–ª–æ—Ö–æ: setState –≤ addListener

```dart
--8<-- "docs/examples/bad/bad-animation-setstate-001.dart"
```

**–ü—Ä–æ–±–ª–µ–º–∞:**
- –í–µ—Å—å –≤–∏–¥–∂–µ—Ç –ø–µ—Ä–µ—Å–æ–±–∏—Ä–∞–µ—Ç—Å—è 60 —Ä–∞–∑ –≤ —Å–µ–∫—É–Ω–¥—É
- Container, Center, Text - –≤—Å—ë –ø–µ—Ä–µ—Å–æ–∑–¥–∞–µ—Ç—Å—è
- –ù–µ—ç—Ñ—Ñ–µ–∫—Ç–∏–≤–Ω–æ–µ –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏–µ —Ä–µ—Å—É—Ä—Å–æ–≤
- –í–æ–∑–º–æ–∂–Ω—ã –ø—Ä–æ–ø—É—Å–∫–∏ –∫–∞–¥—Ä–æ–≤ –Ω–∞ —Å–ª–æ–∂–Ω—ã—Ö UI

## ‚úÖ –•–æ—Ä–æ—à–æ: AnimatedBuilder

```dart
--8<-- "docs/examples/good/good-animation-builder-001.dart"
```

**–ü—Ä–µ–∏–º—É—â–µ—Å—Ç–≤–∞:**
- –ü–µ—Ä–µ—Å–æ–±–∏—Ä–∞–µ—Ç—Å—è —Ç–æ–ª—å–∫–æ Transform.rotate
- –î–æ—á–µ—Ä–Ω–∏–π –≤–∏–¥–∂–µ—Ç (Container) —Å–æ–∑–¥–∞–µ—Ç—Å—è –æ–¥–∏–Ω —Ä–∞–∑
- –û–ø—Ç–∏–º–∏–∑–∏—Ä–æ–≤–∞–Ω–æ –Ω–∞ —É—Ä–æ–≤–Ω–µ —Ñ—Ä–µ–π–º–≤–æ—Ä–∫–∞
- –ù–∏–∫–∞–∫–∏—Ö –Ω–µ–Ω—É–∂–Ω—ã—Ö rebuilds

## ‚úÖ –ï—â–µ –ª—É—á—à–µ: –°–ø–µ—Ü–∏–∞–ª–∏–∑–∏—Ä–æ–≤–∞–Ω–Ω—ã–µ –≤–∏–¥–∂–µ—Ç—ã

```dart
--8<-- "docs/examples/good/good-animation-transition-001.dart"
```

**–°–ø–µ—Ü–∏–∞–ª–∏–∑–∏—Ä–æ–≤–∞–Ω–Ω—ã–µ –≤–∏–¥–∂–µ—Ç—ã:**
- `RotationTransition` - –≤—Ä–∞—â–µ–Ω–∏–µ
- `FadeTransition` - –ø—Ä–æ–∑—Ä–∞—á–Ω–æ—Å—Ç—å
- `ScaleTransition` - –º–∞—Å—à—Ç–∞–±
- `SlideTransition` - —Å–¥–≤–∏–≥
- `PositionedTransition` - –ø–æ–∑–∏—Ü–∏–æ–Ω–∏—Ä–æ–≤–∞–Ω–∏–µ
- `SizeTransition` - —Ä–∞–∑–º–µ—Ä
- `DecoratedBoxTransition` - decoration

–≠—Ç–∏ –≤–∏–¥–∂–µ—Ç—ã **–æ–ø—Ç–∏–º–∏–∑–∏—Ä–æ–≤–∞–Ω—ã** –∏ —Ä–∞–±–æ—Ç–∞—é—Ç –±–µ–∑ rebuilds.

## ‚úÖ CustomPaint —Å –∞–Ω–∏–º–∞—Ü–∏–µ–π

```dart
--8<-- "docs/examples/good/good-animation-custompaint-001.dart"
```

**–ü–æ—á–µ–º—É —ç—Ç–æ –ª—É—á—à–µ:**
- CustomPaint –ø–æ–¥–ø–∏—Å—ã–≤–∞–µ—Ç—Å—è –Ω–∞ Animation –Ω–∞–ø—Ä—è–º—É—é —á–µ—Ä–µ–∑ `super(repaint: ...)`
- –ü–µ—Ä–µ—Ä–∏—Å–æ–≤–∫–∞ –ø—Ä–æ–∏—Å—Ö–æ–¥–∏—Ç –±–µ–∑ rebuild –≤–∏–¥–∂–µ—Ç–∞
- –û–ø—Ç–∏–º–∏–∑–∏—Ä–æ–≤–∞–Ω–æ –Ω–∞ —É—Ä–æ–≤–Ω–µ —Ñ—Ä–µ–π–º–≤–æ—Ä–∫–∞

## ‚úÖ Listenable.merge() –¥–ª—è –º–Ω–æ–∂–µ—Å—Ç–≤–µ–Ω–Ω—ã—Ö –∏—Å—Ç–æ—á–Ω–∏–∫–æ–≤

–ö–æ–≥–¥–∞ –Ω—É–∂–Ω–æ —Å–ª–µ–¥–∏—Ç—å –∑–∞ **–Ω–µ—Å–∫–æ–ª—å–∫–∏–º–∏ –Ω–µ–∑–∞–≤–∏—Å–∏–º—ã–º–∏ –∏—Å—Ç–æ—á–Ω–∏–∫–∞–º–∏** –∏–∑–º–µ–Ω–µ–Ω–∏–π (–Ω–∞–ø—Ä–∏–º–µ—Ä, –∞–Ω–∏–º–∞—Ü–∏—è + –¥–∞–Ω–Ω—ã–µ):

```dart
--8<-- "docs/examples/good/good-custompaint-merge-001.dart"
```

**–ö–∞–∫ —Ä–∞–±–æ—Ç–∞–µ—Ç Listenable.merge():**
- –ü–æ–¥–ø–∏—Å—ã–≤–∞–µ—Ç—Å—è –Ω–∞ **–∫–∞–∂–¥—ã–π** Listenable –≤ —Å–ø–∏—Å–∫–µ
- –ù–æ—Ç–∏—Ñ–∏—Ü–∏—Ä—É–µ—Ç –ø—Ä–∏ –∏–∑–º–µ–Ω–µ–Ω–∏–∏ **–ª—é–±–æ–≥–æ** –∏–∑ –Ω–∏—Ö
- –ò–¥–µ–∞–ª—å–Ω–æ –¥–ª—è —Å–ª—É—á–∞–µ–≤, –∫–æ–≥–¥–∞ –µ—Å—Ç—å –Ω–µ—Å–∫–æ–ª—å–∫–æ –Ω–µ–∑–∞–≤–∏—Å–∏–º—ã—Ö –∏—Å—Ç–æ—á–Ω–∏–∫–æ–≤

**–ö–æ–≥–¥–∞ –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å:**
- ‚úÖ –ê–Ω–∏–º–∞—Ü–∏—è + –∏–∑–º–µ–Ω—è–µ–º—ã–µ –¥–∞–Ω–Ω—ã–µ
- ‚úÖ –ù–µ—Å–∫–æ–ª—å–∫–æ –Ω–µ–∑–∞–≤–∏—Å–∏–º—ã—Ö –∞–Ω–∏–º–∞—Ü–∏–π
- ‚úÖ –ê–Ω–∏–º–∞—Ü–∏—è + –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–∏–π –≤–≤–æ–¥ (ValueNotifier)

## üéØ –§–∞–∑—ã —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞ Flutter

–ü–æ–Ω–∏–º–∞–Ω–∏–µ —Ñ–∞–∑ —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞ –ø–æ–º–æ–≥–∞–µ—Ç –æ—Å–æ–∑–Ω–∞—Ç—å, –ø–æ—á–µ–º—É `repaint`/`relayout` —Ç–∞–∫ —ç—Ñ—Ñ–µ–∫—Ç–∏–≤–Ω—ã.

### –ü–æ–ª–Ω—ã–π —Ü–∏–∫–ª —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞ (—Å setState):

```
1. Build     ‚Üí –ü–µ—Ä–µ—Å–æ–∑–¥–∞–Ω–∏–µ Widget tree
2. Element   ‚Üí –û–±–Ω–æ–≤–ª–µ–Ω–∏–µ Element tree  
3. Layout    ‚Üí –í—ã—á–∏—Å–ª–µ–Ω–∏–µ —Ä–∞–∑–º–µ—Ä–æ–≤ –∏ –ø–æ–∑–∏—Ü–∏–π (RenderObject)
4. Paint     ‚Üí –°–æ–∑–¥–∞–Ω–∏–µ –∏–Ω—Å—Ç—Ä—É–∫—Ü–∏–π –æ—Ç—Ä–∏—Å–æ–≤–∫–∏ (Picture)
5. Composite ‚Üí –û—Ç–ø—Ä–∞–≤–∫–∞ –≤ GPU —á–µ—Ä–µ–∑ Skia
```

### –û–ø—Ç–∏–º–∏–∑–∏—Ä–æ–≤–∞–Ω–Ω—ã–π —Ü–∏–∫–ª (—Å repaint):

```
1. Build     ‚Üí ‚úÖ –ü–†–û–ü–£–°–ö–ê–ï–¢–°–Ø
2. Element   ‚Üí ‚úÖ –ü–†–û–ü–£–°–ö–ê–ï–¢–°–Ø  
3. Layout    ‚Üí ‚úÖ –ü–†–û–ü–£–°–ö–ê–ï–¢–°–Ø
4. Paint     ‚Üí ‚ùå –¢–æ–ª—å–∫–æ –ø–µ—Ä–µ—Ä–∏—Å–æ–≤–∫–∞
5. Composite ‚Üí ‚ùå –û—Ç–ø—Ä–∞–≤–∫–∞ –≤ GPU
```

### –ö–∞–∫ —ç—Ç–æ —Ä–∞–±–æ—Ç–∞–µ—Ç:

**–ü—Ä–∏ –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏–∏ repaint:**
1. –ü—Ä–∏ –∏–∑–º–µ–Ω–µ–Ω–∏–∏ Listenable –ø–æ–º–µ—á–∞–µ—Ç RenderObject –∫–∞–∫ "—Ç—Ä–µ–±—É—é—â–∏–π –ø–µ—Ä–µ—Ä–∏—Å–æ–≤–∫–∏"
2. –ü—Ä–∏ vsync –æ—Ç Flutter Engine –æ–±—Ö–æ–¥–∏—Ç —Ç–æ–ª—å–∫–æ –ø–æ–º–µ—á–µ–Ω–Ω—ã–µ RenderObject
3. –°–æ–±–∏—Ä–∞–µ—Ç Picture (–∏–Ω—Å—Ç—Ä—É–∫—Ü–∏–∏ –æ—Ç—Ä–∏—Å–æ–≤–∫–∏) –∏ –æ—Ç–¥–∞–µ—Ç —Ä–µ–Ω–¥–µ—Ä-–¥–≤–∏–∂–∫—É
4. Skia –æ—Ç—Ä–∏—Å–æ–≤—ã–≤–∞–µ—Ç —à–µ–π–¥–µ—Ä–∞–º–∏ –≤ GPU

**–ü—Ä–µ–∏–º—É—â–µ—Å—Ç–≤–æ:** –ù–∞—á–∏–Ω–∞–µ–º —Å—Ä–∞–∑—É —Å —Ñ–∞–∑—ã Paint, –º–∏–Ω—É—è Build/Element/Layout!

### –°—Ä–∞–≤–Ω–µ–Ω–∏–µ —Å –±—Ä–∞—É–∑–µ—Ä–æ–º:

Flutter –∏—Å–ø–æ–ª—å–∑—É–µ—Ç –ø–æ—Ö–æ–∂–∏–π –º–µ—Ö–∞–Ω–∏–∑–º –Ω–∞ Chrome:

| Flutter | Browser | –ß—Ç–æ –ø—Ä–æ–∏—Å—Ö–æ–¥–∏—Ç |
|---------|---------|----------------|
| Widget/Element | DOM | –°—Ç—Ä—É–∫—Ç—É—Ä–∞ UI |
| Layout | Layout | –í—ã—á–∏—Å–ª–µ–Ω–∏–µ –≥–µ–æ–º–µ—Ç—Ä–∏–∏ |
| Paint | Paint | –°–æ–∑–¥–∞–Ω–∏–µ –∏–Ω—Å—Ç—Ä—É–∫—Ü–∏–π –æ—Ç—Ä–∏—Å–æ–≤–∫–∏ |
| Composite | Composite | –°–±–æ—Ä–∫–∞ —Å–ª–æ–µ–≤ |
| Skia ‚Üí GPU | GPU | –†–µ–Ω–¥–µ—Ä–∏–Ω–≥ |

## ‚úÖ FlowDelegate —Å repaint

–î–ª—è —Å–ª–æ–∂–Ω—ã—Ö —Ç—Ä–∞–Ω—Å—Ñ–æ—Ä–º–∞—Ü–∏–π –º–Ω–æ–∂–µ—Å—Ç–≤–∞ –≤–∏–¥–∂–µ—Ç–æ–≤ –∏—Å–ø–æ–ª—å–∑—É–π—Ç–µ `FlowDelegate` —Å –ø–∞—Ä–∞–º–µ—Ç—Ä–æ–º `repaint`:

```dart
--8<-- "docs/examples/good/good-flow-repaint-001.dart"
```

**–ü—Ä–µ–∏–º—É—â–µ—Å—Ç–≤–∞ FlowDelegate:**
- ‚úÖ –ú–∏–Ω—É–µ—Ç —Ñ–∞–∑—ã Build –∏ Element
- ‚úÖ –í—ã–ø–æ–ª–Ω—è–µ—Ç —Ç–æ–ª—å–∫–æ Layout –∏ Paint
- ‚úÖ –ò–¥–µ–∞–ª—å–Ω–æ –¥–ª—è –∞–Ω–∏–º–∞—Ü–∏–∏ –ø–æ–∑–∏—Ü–∏–π/—Ç—Ä–∞–Ω—Å—Ñ–æ—Ä–º–∞—Ü–∏–π –º–Ω–æ–∂–µ—Å—Ç–≤–∞ –≤–∏–¥–∂–µ—Ç–æ–≤
- ‚úÖ –ë–æ–ª–µ–µ –ø—Ä–æ–∏–∑–≤–æ–¥–∏—Ç–µ–ª—å–Ω–æ —á–µ–º Stack —Å Positioned

**–ö–æ–≥–¥–∞ –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å:**
- –ê–Ω–∏–º–∞—Ü–∏—è –ø–æ–∑–∏—Ü–∏–π –º–Ω–æ–∂–µ—Å—Ç–≤–∞ —ç–ª–µ–º–µ–Ω—Ç–æ–≤
- –°–ª–æ–∂–Ω—ã–µ —Ç—Ä–∞–Ω—Å—Ñ–æ—Ä–º–∞—Ü–∏–∏ (rotation, scale, translate)
- –ö–∞—Å—Ç–æ–º–Ω—ã–µ layout'—ã —Å –∞–Ω–∏–º–∞—Ü–∏–µ–π

## ‚úÖ MultiChildLayoutDelegate —Å relayout

–î–ª—è –∞–Ω–∏–º–∞—Ü–∏–∏ layout'–∞ –∏—Å–ø–æ–ª—å–∑—É–π—Ç–µ `MultiChildLayoutDelegate`:

```dart
--8<-- "docs/examples/good/good-layout-relayout-001.dart"
```

**–ü—Ä–µ–∏–º—É—â–µ—Å—Ç–≤–∞ MultiChildLayoutDelegate:**
- ‚úÖ –ú–∏–Ω—É–µ—Ç —Ñ–∞–∑—É Build –≤–∏–¥–∂–µ—Ç–æ–≤
- ‚úÖ –í—ã–ø–æ–ª–Ω—è–µ—Ç Layout –∏ Paint
- ‚úÖ –ò–¥–µ–∞–ª—å–Ω–æ –¥–ª—è –∞–Ω–∏–º–∞—Ü–∏–∏ —Ä–∞–∑–º–µ—Ä–æ–≤/–ø–æ–∑–∏—Ü–∏–π —Å–µ–∫—Ü–∏–π
- ‚úÖ –ë–æ–ª–µ–µ –≥–∏–±–∫–∏–π —á–µ–º Column/Row –¥–ª—è –∫–∞—Å—Ç–æ–º–Ω—ã—Ö layout'–æ–≤

**–ö–æ–≥–¥–∞ –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å:**
- –ê–Ω–∏–º–∞—Ü–∏—è —Ä–∞–∑–º–µ—Ä–æ–≤ —Å–µ–∫—Ü–∏–π UI
- –î–∏–Ω–∞–º–∏—á–µ—Å–∫–∏–µ layout'—ã —Å –∏–∑–º–µ–Ω—è—é—â–∏–º–∏—Å—è –ø—Ä–æ–ø–æ—Ä—Ü–∏—è–º–∏
- –°–ª–æ–∂–Ω–∞—è –≥–µ–æ–º–µ—Ç—Ä–∏—è —Å –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç—è–º–∏ –º–µ–∂–¥—É —ç–ª–µ–º–µ–Ω—Ç–∞–º–∏

## RepaintBoundary –¥–ª—è –∏–∑–æ–ª—è—Ü–∏–∏

–ò—Å–ø–æ–ª—å–∑—É–π—Ç–µ `RepaintBoundary` –¥–ª—è –∏–∑–æ–ª—è—Ü–∏–∏ –∞–Ω–∏–º–∏—Ä–æ–≤–∞–Ω–Ω—ã—Ö —á–∞—Å—Ç–µ–π UI:

```dart
// ‚ùå –ü–õ–û–•–û - –∞–Ω–∏–º–∞—Ü–∏—è –ø–µ—Ä–µ—Ä–∏—Å–æ–≤—ã–≤–∞–µ—Ç –≤—Å—ë –¥–µ—Ä–µ–≤–æ
Column(
  children: [
    ExpensiveWidget(),
    AnimatedWidget(), // –í—ã–∑—ã–≤–∞–µ—Ç repaint –≤—Å–µ–π Column!
    AnotherExpensiveWidget(),
  ],
)

// ‚úÖ –•–û–†–û–®–û - –∏–∑–æ–ª–∏—Ä—É–µ–º –∞–Ω–∏–º–∞—Ü–∏—é
Column(
  children: [
    ExpensiveWidget(),
    RepaintBoundary(
      child: AnimatedWidget(), // –ò–∑–æ–ª–∏—Ä–æ–≤–∞–Ω!
    ),
    AnotherExpensiveWidget(),
  ],
)
```

### –ö–æ–≥–¥–∞ –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å RepaintBoundary:

- ‚úÖ –í–æ–∫—Ä—É–≥ –∞–Ω–∏–º–∏—Ä–æ–≤–∞–Ω–Ω—ã—Ö –≤–∏–¥–∂–µ—Ç–æ–≤
- ‚úÖ –í–æ–∫—Ä—É–≥ CustomPaint
- ‚úÖ –í–æ–∫—Ä—É–≥ —Å–ª–æ–∂–Ω—ã—Ö –≤–∏–¥–∂–µ—Ç–æ–≤, –∫–æ—Ç–æ—Ä—ã–µ —Ä–µ–¥–∫–æ –º–µ–Ω—è—é—Ç—Å—è
- ‚úÖ –í —Å–ø–∏—Å–∫–∞—Ö –¥–ª—è –∏–∑–æ–ª—è—Ü–∏–∏ —ç–ª–µ–º–µ–Ω—Ç–æ–≤

### –ö–æ–≥–¥–∞ –ù–ï –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å:

- ‚ùå –í–µ–∑–¥–µ –ø–æ–¥—Ä—è–¥ (–µ—Å—Ç—å overhead)
- ‚ùå –ù–∞ –º–µ–ª–∫–∏—Ö –≤–∏–¥–∂–µ—Ç–∞—Ö
- ‚ùå –ù–∞ –≤–∏–¥–∂–µ—Ç–∞—Ö, –∫–æ—Ç–æ—Ä—ã–µ –∏ —Ç–∞–∫ —á–∞—Å—Ç–æ –ø–µ—Ä–µ—Ä–∏—Å–æ–≤—ã–≤–∞—é—Ç—Å—è

## –°—Ä–∞–≤–Ω–µ–Ω–∏–µ –ø–æ–¥—Ö–æ–¥–æ–≤

### –¢–∞–±–ª–∏—Ü–∞ –ø—Ä–æ–∏–∑–≤–æ–¥–∏—Ç–µ–ª—å–Ω–æ—Å—Ç–∏ –ø–æ —Ñ–∞–∑–∞–º —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞:

| –ü–æ–¥—Ö–æ–¥ | Build | Element | Layout | Paint | –ü—Ä–æ–∏–∑–≤–æ–¥–∏—Ç–µ–ª—å–Ω–æ—Å—Ç—å |
|--------|-------|---------|--------|-------|-------------------|
| `setState + addListener` | ‚ùå –î–∞ | ‚ùå –î–∞ | ‚ùå –î–∞ | ‚ùå –î–∞ | üíÄ –û—á–µ–Ω—å –ø–ª–æ—Ö–æ |
| `AnimatedBuilder` | ‚ö†Ô∏è –ß–∞—Å—Ç–∏—á–Ω–æ | ‚ö†Ô∏è –ß–∞—Å—Ç–∏—á–Ω–æ | ‚ö†Ô∏è –ß–∞—Å—Ç–∏—á–Ω–æ | ‚ùå –î–∞ | üü° –°—Ä–µ–¥–Ω–µ |
| –°–ø–µ—Ü–∏–∞–ª–∏–∑–∏—Ä–æ–≤–∞–Ω–Ω—ã–µ –≤–∏–¥–∂–µ—Ç—ã | ‚úÖ –ù–µ—Ç | ‚úÖ –ù–µ—Ç | ‚úÖ –ù–µ—Ç | ‚ùå –î–∞ | üü¢ –û—Ç–ª–∏—á–Ω–æ |
| `CustomPaint + repaint` | ‚úÖ –ù–µ—Ç | ‚úÖ –ù–µ—Ç | ‚úÖ –ù–µ—Ç | ‚ùå –î–∞ | üü¢ –û—Ç–ª–∏—á–Ω–æ |
| `FlowDelegate + repaint` | ‚úÖ –ù–µ—Ç | ‚úÖ –ù–µ—Ç | ‚ùå –î–∞ | ‚ùå –î–∞ | üü¢ –û—Ç–ª–∏—á–Ω–æ |
| `MultiChildLayoutDelegate + relayout` | ‚úÖ –ù–µ—Ç | ‚úÖ –ù–µ—Ç | ‚ùå –î–∞ | ‚ùå –î–∞ | üü° –•–æ—Ä–æ—à–æ |

### –ö–æ–≥–¥–∞ —á—Ç–æ –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å:

| –ü–æ–¥—Ö–æ–¥ | –ö–æ–≥–¥–∞ –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å |
|--------|-------------------|
| `setState + addListener` | ‚ùå **–ù–∏–∫–æ–≥–¥–∞** |
| `AnimatedBuilder` | ‚úÖ –£–Ω–∏–≤–µ—Ä—Å–∞–ª—å–Ω–æ–µ —Ä–µ—à–µ–Ω–∏–µ –¥–ª—è –±–æ–ª—å—à–∏–Ω—Å—Ç–≤–∞ –∞–Ω–∏–º–∞—Ü–∏–π |
| –°–ø–µ—Ü–∏–∞–ª–∏–∑–∏—Ä–æ–≤–∞–Ω–Ω—ã–µ –≤–∏–¥–∂–µ—Ç—ã | ‚úÖ –ü—Ä–æ—Å—Ç—ã–µ –∞–Ω–∏–º–∞—Ü–∏–∏ (rotation, fade, scale, slide) |
| `CustomPaint + repaint` | ‚úÖ –ö–∞—Å—Ç–æ–º–Ω–∞—è –æ—Ç—Ä–∏—Å–æ–≤–∫–∞, canvas, –≥—Ä–∞—Ñ–∏–∫–∏ |
| `FlowDelegate + repaint` | ‚úÖ –ê–Ω–∏–º–∞—Ü–∏—è –ø–æ–∑–∏—Ü–∏–π/—Ç—Ä–∞–Ω—Å—Ñ–æ—Ä–º–∞—Ü–∏–π –º–Ω–æ–∂–µ—Å—Ç–≤–∞ –≤–∏–¥–∂–µ—Ç–æ–≤ |
| `MultiChildLayoutDelegate + relayout` | ‚úÖ –ê–Ω–∏–º–∞—Ü–∏—è —Ä–∞–∑–º–µ—Ä–æ–≤/–≥–µ–æ–º–µ—Ç—Ä–∏–∏ —Å–µ–∫—Ü–∏–π UI |

## –ü—Ä–∞–≤–∏–ª–æ —Ä–∞–±–æ—Ç—ã —Å AnimationController

```dart
// ‚ùå –ù–ò–ö–û–ì–î–ê –Ω–µ –¥–µ–ª–∞–π—Ç–µ —Ç–∞–∫:
_controller.addListener(() {
  setState(() {});
});

// ‚úÖ –ò—Å–ø–æ–ª—å–∑—É–π—Ç–µ AnimatedBuilder:
AnimatedBuilder(
  animation: _controller,
  builder: (context, child) => ...,
)

// ‚úÖ –ò–ª–∏ —Å–ø–µ—Ü–∏–∞–ª–∏–∑–∏—Ä–æ–≤–∞–Ω–Ω—ã–µ –≤–∏–¥–∂–µ—Ç—ã:
RotationTransition(turns: _controller, child: ...)

// ‚úÖ –ò–ª–∏ CustomPaint —Å repaint:
CustomPaint(painter: MyPainter(animation: _controller))
class MyPainter extends CustomPainter {
  MyPainter({required this.animation}) : super(repaint: animation);
  final Animation animation;
}
```

## Checklist

- [ ] –ù–µ—Ç `addListener(() => setState())` –¥–ª—è AnimationController
- [ ] –ò—Å–ø–æ–ª—å–∑—É–µ—Ç—Å—è `AnimatedBuilder` –∏–ª–∏ —Å–ø–µ—Ü–∏–∞–ª–∏–∑–∏—Ä–æ–≤–∞–Ω–Ω—ã–µ –≤–∏–¥–∂–µ—Ç—ã
- [ ] CustomPaint –ø–æ–ª—É—á–∞–µ—Ç Animation —á–µ—Ä–µ–∑ `super(repaint: animation)`
- [ ] –ò—Å–ø–æ–ª—å–∑—É–µ—Ç—Å—è `Listenable.merge()` –¥–ª—è –º–Ω–æ–∂–µ—Å—Ç–≤–µ–Ω–Ω—ã—Ö –∏—Å—Ç–æ—á–Ω–∏–∫–æ–≤ –∏–∑–º–µ–Ω–µ–Ω–∏–π
- [ ] FlowDelegate –∏—Å–ø–æ–ª—å–∑—É–µ—Ç `super(repaint: animation)` –≤–º–µ—Å—Ç–æ setState
- [ ] MultiChildLayoutDelegate –∏—Å–ø–æ–ª—å–∑—É–µ—Ç `super(relayout: animation)` 
- [ ] –ê–Ω–∏–º–∏—Ä–æ–≤–∞–Ω–Ω—ã–µ –≤–∏–¥–∂–µ—Ç—ã –æ–±–µ—Ä–Ω—É—Ç—ã –≤ `RepaintBoundary` –ø—Ä–∏ –Ω–µ–æ–±—Ö–æ–¥–∏–º–æ—Å—Ç–∏
- [ ] –î–æ—á–µ—Ä–Ω–∏–µ –≤–∏–¥–∂–µ—Ç—ã –ø–µ—Ä–µ–¥–∞—é—Ç—Å—è —á–µ—Ä–µ–∑ `child` –ø–∞—Ä–∞–º–µ—Ç—Ä –¥–ª—è –ø–µ—Ä–µ–∏—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏—è
- [ ] –ü–æ–Ω—è—Ç–Ω—ã —Ñ–∞–∑—ã —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞ (Build ‚Üí Element ‚Üí Layout ‚Üí Paint ‚Üí Composite)

## –°–º. —Ç–∞–∫–∂–µ

- [PERF-COLLECTIONS-IDENTICAL](./PERF-COLLECTIONS-IDENTICAL.md) - —ç—Ñ—Ñ–µ–∫—Ç–∏–≤–Ω–æ–µ —Å—Ä–∞–≤–Ω–µ–Ω–∏–µ –≤ shouldRepaint
- [Flutter Animation Best Practices](https://docs.flutter.dev/development/ui/animations/tutorial)

