---
id: DART-DATA-FACTORY
title: "Используй factory для сценариев с отсутствующими/генерируемыми полями"
status: stable
severity: warning
category: dart/data-class
tags: [factory, constructor]
linter_rule:
coverage: ai
bad_snippet: bad-factory-001.dart
good_snippet: good-factory-001.dart
references: []
ai_hint: >
  Detect repeated creation patterns with missing fields and suggest factory constructor
  that explicitly fills null/defaults and delegates to the strict constructor.
---

## Пояснение

При создании объектов с отсутствующими или генерируемыми полями следует использовать factory конструкторы. Это обеспечивает явность в коде и делегирование к основному строгому конструктору.

## Основные принципы

1. **Используем factory конструктор** для сценариев с отсутствующими полями
2. **Явно указываем `null`/значения по умолчанию** для недостающих полей
3. **Делегируем к основному строгому конструктору**
4. **Избегаем дублирования логики** создания объектов

## Плохо

```dart
--8<-- "docs/examples/bad/bad-factory-001.dart"
```

**Проблемы:**
- Дублирование логики создания объектов
- Неявные значения по умолчанию
- Отсутствие делегирования к основному конструктору
- Пустые строки вместо `null` для отсутствующих полей

### Пример с nullable полями

```dart
--8<-- "docs/examples/bad/bad-thing-location-model.dart"
```

**Дополнительные проблемы:**
- Nullable поля без явного указания значений по умолчанию
- Изменяемые поля вместо `final`
- Дублирование логики инициализации в разных конструкторах

## Хорошо

```dart
--8<-- "docs/examples/good/good-factory-001.dart"
```

**Преимущества:**
- Явность: каждый factory показывает, какие поля отсутствуют
- Делегирование: все factory используют основной конструктор
- Читаемость: понятно, откуда берутся значения
- Гибкость: разные сценарии создания через отдельные factory

### Пример с nullable полями

```dart
--8<-- "docs/examples/good/good-thing-location-model.dart"
```

**Дополнительные преимущества:**
- Строгий основной конструктор с `required` для обязательных полей
- Factory для разных сценариев создания (временные данные, минимальные данные, JSON)
- Иммутабельность через `final` поля
- Явные значения по умолчанию для всех nullable полей

## Примеры использования

### Factory для создания с временными данными

```dart
--8<-- "docs/examples/good/good-factory-002.dart"
```

### Factory с валидацией

```dart
--8<-- "docs/examples/good/good-factory-003.dart"
```

### Factory для модели с nullable полями

```dart
--8<-- "docs/examples/good/good-thing-location-model.dart"
```

**Ключевые моменты:**
- Основной конструктор использует `required` для обязательных полей
- Factory конструкторы явно указывают значения по умолчанию
- Все factory делегируют к основному конструктору
- Поля сделаны `final` для иммутабельности

## Проверка правила

### Что проверяется

- [ ] Для сценариев с отсутствующими полями используется factory конструктор
- [ ] Factory явно указывает значения по умолчанию для недостающих полей
- [ ] Все factory делегируют к основному строгому конструктору
- [ ] Отсутствует дублирование логики создания объектов
- [ ] Factory имеют понятные названия, отражающие сценарий создания
- [ ] Основной конструктор использует `required` для обязательных полей
- [ ] Nullable поля имеют явные значения по умолчанию в factory
- [ ] Поля сделаны `final` для иммутабельности (где это возможно)

### Автоматическая проверка

Это правило проверяется ИИ-агентами при анализе кода. Убедитесь, что все сценарии создания объектов с отсутствующими полями используют factory конструкторы.

## Связанные правила

- [DART-DATA-REQ-CTOR](DART-DATA-REQ-CTOR.md) - Строгие конструкторы
- [DART-DATA-JSON](DART-DATA-JSON.md) - JSON сериализация
