---
id: DART-DATA-FACTORY
title: "Используй factory для сценариев с отсутствующими/генерируемыми полями"
status: stable
severity: warning
category: dart/data-class
tags: [factory, constructor]
linter_rule:
coverage: ai
bad_snippet: bad-factory-001.dart
good_snippet: good-factory-001.dart
references: []
ai_hint: >
  Detect repeated creation patterns with missing fields and suggest factory constructor
  that explicitly fills null/defaults and delegates to the strict constructor.
---

## Пояснение

При создании объектов с отсутствующими или генерируемыми полями следует использовать factory конструкторы. Это обеспечивает явность в коде и делегирование к основному строгому конструктору.

## Основные принципы

1. **Используем factory конструктор** для сценариев с отсутствующими полями
2. **Явно указываем `null`/значения по умолчанию** для недостающих полей
3. **Делегируем к основному строгому конструктору**
4. **Избегаем дублирования логики** создания объектов

## Плохо

```dart
--8<-- "docs/examples/bad/bad-factory-001.dart"
```

**Проблемы:**
- Дублирование логики создания объектов
- Неявные значения по умолчанию
- Отсутствие делегирования к основному конструктору
- Пустые строки вместо `null` для отсутствующих полей

## Хорошо

```dart
--8<-- "docs/examples/good/good-factory-001.dart"
```

**Преимущества:**
- Явность: каждый factory показывает, какие поля отсутствуют
- Делегирование: все factory используют основной конструктор
- Читаемость: понятно, откуда берутся значения
- Гибкость: разные сценарии создания через отдельные factory

## Примеры использования

### Factory для создания с временными данными

```dart
--8<-- "docs/examples/good/good-factory-002.dart"
```

### Factory с валидацией

```dart
--8<-- "docs/examples/good/good-factory-003.dart"
```

## Проверка правила

### Что проверяется

- [ ] Для сценариев с отсутствующими полями используется factory конструктор
- [ ] Factory явно указывает значения по умолчанию для недостающих полей
- [ ] Все factory делегируют к основному строгому конструктору
- [ ] Отсутствует дублирование логики создания объектов
- [ ] Factory имеют понятные названия, отражающие сценарий создания

### Автоматическая проверка

Это правило проверяется ИИ-агентами при анализе кода. Убедитесь, что все сценарии создания объектов с отсутствующими полями используют factory конструкторы.

## Связанные правила

- [DART-DATA-REQ-CTOR](DART-DATA-REQ-CTOR.md) - Строгие конструкторы
- [DART-DATA-JSON](DART-DATA-JSON.md) - JSON сериализация
