# –†–µ–ø–æ–∑–∏—Ç–æ—Ä–∏–π —Å—Ç–∞—Ç–∏—á–µ–Ω: –±–µ–∑ init/setup, –µ–¥–∏–Ω—ã–π Future-API –∏ –∫–æ—Ä—Ä–µ–∫—Ç–Ω—ã–µ –≥–ª–∞–≥–æ–ª—ã

**ID:** `ARCH-REPO-STATELESS`  
**–°—Ç–∞—Ç—É—Å:** –°—Ç–∞–±–∏–ª—å–Ω—ã–π  
**–í–∞–∂–Ω–æ—Å—Ç—å:** –û—à–∏–±–∫–∞  
**–ö–∞—Ç–µ–≥–æ—Ä–∏—è:** –ê—Ä—Ö–∏—Ç–µ–∫—Ç—É—Ä–∞ / –†–µ–ø–æ–∑–∏—Ç–æ—Ä–∏–π  

## üè∑Ô∏è –¢–µ–≥–∏
- clean-architecture
- repository
- stateless
- idempotency
- boundaries
- naming
- api-shape
- di

**–í–µ—Ä—Å–∏—è:** 3  
**–í–ª–∞–¥–µ–ª—å—Ü—ã:** @arch-bot, @team-data

---

## üìç –ì–¥–µ –ø—Ä–æ–≤–µ—Ä—è–µ–º

**–í–∫–ª—é—á–∞–µ–º:**
- `lib/**/repository/**`
- `lib/**/data/**`
- `lib/**/infra/**`
- `lib/**/**repository**.dart`

**–ò—Å–∫–ª—é—á–∞–µ–º:**
- `lib/**/ui/**`
- `lib/**/api/**`
- `lib/**/presentation/**`
- `lib/**/widget/**`
- `lib/**/view/**`
- `lib/**/bloc/**`
- `lib/**/cubit/**`
- `test/**`
- `integration_test/**`

---

## üîç –î–µ—Ç–µ–∫—Ç–æ—Ä—ã –Ω–∞—Ä—É—à–µ–Ω–∏–π

### –†–∞–∑—Ä–µ—à–µ–Ω–Ω—ã–µ –∏–º–ø–æ—Ä—Ç—ã
- `dart:async`

### –ü–∞—Ç—Ç–µ—Ä–Ω—ã –¥–ª—è –ø—Ä–æ–≤–µ—Ä–∫–∏

#### 1. ‚ùå –ó–∞–ø—Ä–µ—â–µ–Ω—ã init/setup –º–µ—Ç–æ–¥—ã
**–û–ø–∏—Å–∞–Ω–∏–µ:** –í —Ä–µ–ø–æ–∑–∏—Ç–æ—Ä–∏–∏ –Ω–∞–π–¥–µ–Ω init/setup/start/bootstrap/preload/configure/loadCache/warmup/open/connect ‚Äî —Ä–µ–ø–æ–∑–∏—Ç–æ—Ä–∏–π –¥–æ–ª–∂–µ–Ω –±—ã—Ç—å –±–µ—Å—Å—Ç–µ–π—Ç–Ω—ã–º.

**–†–µ–≥—É–ª—è—Ä–∫–∞:** `(?:class|abstract\s+class)\s+\w+Repository(?:I)?\b[\s\S]*?\b(?:Future\s*<[^>]+>|Future|void)\s+(?:init(?:[A-Z_].*)?|initialize(?:[A-Z_].*)?|setup(?:[A-Z_].*)?|start(?:[A-Z_].*)?|bootstrap(?:[A-Z_].*)?|preload(?:[A-Z_].*)?|configure(?:[A-Z_].*)?|loadCache(?:[A-Z_].*)?|warmup(?:[A-Z_].*)?|open(?:[A-Z_].*)?|connect(?:[A-Z_].*)?)\s*\(`

#### 2. ‚ùå –ó–∞–ø—Ä–µ—â–µ–Ω—ã Stream –≤ –ø—É–±–ª–∏—á–Ω–æ–º API
**–û–ø–∏—Å–∞–Ω–∏–µ:** –ü—É–±–ª–∏—á–Ω—ã–π –º–µ—Ç–æ–¥ —Ä–µ–ø–æ–∑–∏—Ç–æ—Ä–∏—è –≤–æ–∑–≤—Ä–∞—â–∞–µ—Ç Stream ‚Äî –ø—É–±–ª–∏—á–Ω—ã–π API –¥–æ–ª–∂–µ–Ω –±—ã—Ç—å Future-based.

**–†–µ–≥—É–ª—è—Ä–∫–∞:** `(?:class|abstract\s+class)\s+\w+Repository(?:I)?\b[\s\S]*?\bStream\s*<[^>]+>\s+(?!_)[a-z][A-Za-z0-9_]*\s*\(`

#### 3. ‚ùå –ó–∞–ø—Ä–µ—â–µ–Ω public void (–∫—Ä–æ–º–µ dispose)
**–û–ø–∏—Å–∞–Ω–∏–µ:** –ü—É–±–ª–∏—á–Ω—ã–π –º–µ—Ç–æ–¥ —Ä–µ–ø–æ–∑–∏—Ç–æ—Ä–∏—è –≤–æ–∑–≤—Ä–∞—â–∞–µ—Ç void (–∫—Ä–æ–º–µ dispose) ‚Äî –∏—Å–ø–æ–ª—å–∑—É–π—Ç–µ Future<...>.

**–†–µ–≥—É–ª—è—Ä–∫–∞:** `(?:class|abstract\s+class)\s+\w+Repository(?:I)?\b[\s\S]*?\bvoid\s+(?!dispose\b)(?!_)[a-z][A-Za-z0-9_]*\s*\(`

#### 4. ‚ùå –ó–∞–ø—Ä–µ—â–µ–Ω –Ω–µ-Future –≤–æ–∑–≤—Ä–∞—Ç
**–û–ø–∏—Å–∞–Ω–∏–µ:** –ü—É–±–ª–∏—á–Ω—ã–π –º–µ—Ç–æ–¥ —Ä–µ–ø–æ–∑–∏—Ç–æ—Ä–∏—è –Ω–µ –≤–æ–∑–≤—Ä–∞—â–∞–µ—Ç Future ‚Äî –ø—É–±–ª–∏—á–Ω—ã–π API –¥–æ–ª–∂–µ–Ω –±—ã—Ç—å –∞—Å–∏–Ω—Ö—Ä–æ–Ω–Ω—ã–º.

**–†–µ–≥—É–ª—è—Ä–∫–∞:** `(?:class|abstract\s+class)\s+\w+Repository(?:I)?\b[\s\S]*?\b(?!Future\s*<[^>]+>|Future\b|void\b|Stream\s*<[^>]+>|factory\b|const\b|static\b)[A-Za-z_][A-Za-z0-9_<>?,\[\]\s]*\s+(?!_)[a-z][A-Za-z0-9_]*\s*\(`

#### 5. ‚ùå –ù–∞—Ä—É—à–µ–Ω–∏–µ —Å–æ–≥–ª–∞—à–µ–Ω–∏—è –æ–± –∏–º–µ–Ω–∞—Ö
**–û–ø–∏—Å–∞–Ω–∏–µ:** –ü—É–±–ª–∏—á–Ω—ã–µ –º–µ—Ç–æ–¥—ã —Ä–µ–ø–æ–∑–∏—Ç–æ—Ä–∏—è –¥–æ–ª–∂–Ω—ã –Ω–∞—á–∏–Ω–∞—Ç—å—Å—è —Å fetch/create/update/delete (–Ω–∞–ø—Ä–∏–º–µ—Ä, get*/preload*/load*/list* ‚Äî –∑–∞–ø—Ä–µ—â–µ–Ω—ã).

**–†–µ–≥—É–ª—è—Ä–∫–∞:** `(?:class|abstract\s+class)\s+\w+Repository(?:I)?\b[\s\S]*?\b(?:Future\s*<[^>]+>|Future|void|[A-Za-z_][A-Za-z0-9_<>,\s\[\]?]*)\s+(?!_)(?!dispose\b)(?!(?:fetch|create|update|delete)\b)[a-z][A-Za-z0-9_]*\s*\(`

---

## üí¨ –°–æ–æ–±—â–µ–Ω–∏–µ –æ–± –æ—à–∏–±–∫–µ

–†–µ–ø–æ–∑–∏—Ç–æ—Ä–∏–π –¥–æ–ª–∂–µ–Ω –±—ã—Ç—å –±–µ—Å—Å—Ç–µ–π—Ç–Ω—ã–º –∏ –∏–º–µ—Ç—å –µ–¥–∏–Ω—ã–π Future-based API. –ü—É–±–ª–∏—á–Ω—ã–µ –º–µ—Ç–æ–¥—ã –¥–æ–ª–∂–Ω—ã –Ω–∞—á–∏–Ω–∞—Ç—å—Å—è —Å fetch/create/update/delete. –ó–∞–ø—Ä–µ—â–µ–Ω—ã init/setup/start/bootstrap/preload –∏ –ø—Ä., Stream –≤ –ø—É–±–ª–∏—á–Ω–æ–º API, public void (–∫—Ä–æ–º–µ dispose), –∏ –ª—é–±—ã–µ public-–º–µ—Ç–æ–¥—ã —Å –∏–º–µ–Ω–∞–º–∏ –≤–Ω–µ allowed-–≥–ª–∞–≥–æ–ª–æ–≤.

---

## üîß –ê–≤—Ç–æ–∏—Å–ø—Ä–∞–≤–ª–µ–Ω–∏–µ

**–°—Ç—Ä–æ–∏—Ç–µ–ª—å –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏–π:** –ú–∏–Ω–∏–º–∞–ª—å–Ω—ã–π

**–ü—Ä–µ–¥–ª–æ–∂–µ–Ω–∏–µ:** –ü–µ—Ä–µ–∏–º–µ–Ω—É–π—Ç–µ –º–µ—Ç–æ–¥—ã –Ω–∞ fetch/create/update/delete (–Ω–∞–ø—Ä–∏–º–µ—Ä, getDistinctValues ‚Üí fetchDistinctValues). –£–¥–∞–ª–∏—Ç–µ/–ø–µ—Ä–µ–Ω–µ—Å–∏—Ç–µ preload/init/setup –≤ DI/–∫–æ–Ω—Å—Ç—Ä—É–∫—Ç–æ—Ä —Ä–µ–∞–ª–∏–∑–∞—Ü–∏–∏. –ü–æ—Ç–æ–∫–∏ –≤—ã–Ω–µ—Å–∏—Ç–µ –≤ watcher-—Å–µ—Ä–≤–∏—Å, –ø—É–±–ª–∏—á–Ω—ã–µ –º–µ—Ç–æ–¥—ã —Å–¥–µ–ª–∞–π—Ç–µ Future<...>.

---

## üß™ –ü—Ä–∏–º–µ—Ä—ã –∏ —Ç–µ—Å—Ç—ã

### ‚ùå –î–æ–ª–∂–Ω–æ —Ñ–ª–∞–≥–∏—Ä–æ–≤–∞—Ç—å—Å—è

#### –ü—Ä–∏–º–µ—Ä 1: –ù–µ–ø—Ä–∞–≤–∏–ª—å–Ω—ã–µ –∏–º–µ–Ω–∞ –º–µ—Ç–æ–¥–æ–≤ –∏ preload
```dart:lib/data/filters/repository/filter_repository.dart
import 'package:injectable/injectable.dart';
import '../network/filter_data_api.dart';

@LazySingleton(as: FilterRepositoryI)
class FilterRepository implements FilterRepositoryI {
  FilterRepository({required FilterDataApiI api}) : _api = api;
  final FilterDataApiI _api;

  @override
  Future<List<String>> getDistinctValues(String field) { // ‚ùå –∏–º—è –Ω–µ —Å fetch/create/update/delete
    return _api.fetchDistinctValuesForCurrentUser(field: field);
  }

  @override
  Future<Map<String, List<String>>> preload(Iterable<String> fields) async { // ‚ùå preload –∑–∞–ø—Ä–µ—â—ë–Ω
    final entries = await Future.wait(
      fields.map((f) async {
        final values = await _api.fetchDistinctValuesForCurrentUser(field: f);
        return MapEntry(f, values);
      }),
    );
    return Map<String, List<String>>.fromEntries(entries);
  }
}

abstract class FilterRepositoryI {
  Future<List<String>> getDistinctValues(String field); // ‚ùå –∏–º—è
  Future<Map<String, List<String>>> preload(Iterable<String> fields); // ‚ùå preload
}
```

#### –ü—Ä–∏–º–µ—Ä 2: –†–∞–∑–ª–∏—á–Ω—ã–µ –Ω–∞—Ä—É—à–µ–Ω–∏—è
```dart:lib/data/common/repository/preload_repo.dart
class SomeRepository {
  Future<void> preloadCache() async {} // ‚ùå init-like/preload
  Stream<int> watch() => const Stream.empty(); // ‚ùå Stream –≤ –ø—É–±–ª–∏—á–Ω–æ–º API
  void clear() {} // ‚ùå public void
}
```

### ‚úÖ –î–æ–ª–∂–Ω–æ –ø—Ä–æ—Ö–æ–¥–∏—Ç—å

#### –ü—Ä–∏–º–µ—Ä: –ü—Ä–∞–≤–∏–ª—å–Ω–∞—è —Ä–µ–∞–ª–∏–∑–∞—Ü–∏—è
```dart:lib/data/filters/repository/filter_repository_ok.dart
import 'package:injectable/injectable.dart';
import '../network/filter_data_api.dart';

abstract class FilterRepositoryI {
  Future<List<String>> fetchDistinctValues(String field);
  Future<Map<String, List<String>>> fetchDistinctValuesByFields(Iterable<String> fields);
}

@LazySingleton(as: FilterRepositoryI)
class FilterRepository implements FilterRepositoryI {
  FilterRepository({required FilterDataApiI api}) : _api = api;
  final FilterDataApiI _api;

  @override
  Future<List<String>> fetchDistinctValues(String field) {
    return _api.fetchDistinctValuesForCurrentUser(field: field);
  }

  @override
  Future<Map<String, List<String>>> fetchDistinctValuesByFields(Iterable<String> fields) async {
    final entries = await Future.wait(
      fields.map((f) async {
        final values = await _api.fetchDistinctValuesForCurrentUser(field: f);
        return MapEntry(f, values);
      }),
    );
    return Map<String, List<String>>.fromEntries(entries);
  }

  void dispose() {} // –¥–æ–ø—É—Å—Ç–∏–º–æ, –µ—Å–ª–∏ —Ä–∞–∑–º–µ—â—ë–Ω –ø–æ—Å–ª–µ–¥–Ω–∏–º
}
```

---

## üìö –ü–æ—è—Å–Ω–µ–Ω–∏–µ

–†–µ–ø–æ–∑–∏—Ç–æ—Ä–∏–π ‚Äî —Ç–æ–Ω–∫–∞—è, **–±–µ—Å—Å—Ç–µ–π—Ç–Ω–∞—è** –∞–±—Å—Ç—Ä–∞–∫—Ü–∏—è –Ω–∞–¥ –¥–∞–Ω–Ω—ã–º–∏:

### ‚úÖ –ß—Ç–æ –¥–æ–ª–∂–Ω–æ –±—ã—Ç—å:
- **–ù–∏–∫–∞–∫–∏—Ö `init()/setup()`** –∏ —Ç.–ø.; —Ä–µ–∑—É–ª—å—Ç–∞—Ç –º–µ—Ç–æ–¥–æ–≤ –Ω–µ –∑–∞–≤–∏—Å–∏—Ç –æ—Ç –ø–æ—Ä—è–¥–∫–∞ –≤—ã–∑–æ–≤–æ–≤
- **–í–æ–∑–≤—Ä–∞—â–∞–µ—Ç `Future<...>`** –≤ –ø—É–±–ª–∏—á–Ω–æ–º API. –ï—Å–ª–∏ –∏—Å—Ç–æ—á–Ω–∏–∫ ‚Äî `Stream`, –∫–æ–Ω–≤–µ—Ä—Ç–∏—Ä—É–π—Ç–µ –≤–Ω—É—Ç—Ä–∏ (`.first`, —Å–±–æ—Ä–∫–∞ –≤ —Å–ø–∏—Å–æ–∫ –∏ —Ç.–ø.)
- **–ï–¥–∏–Ω—ã–π –∑–∞–ø—Ä–æ—Å `fetch(FetchFilter filter)`** –≤–º–µ—Å—Ç–æ –≤–∞—Ä–∏–∞—Ü–∏–π `fetchXOnce/fetchMyY`. –§–∏–ª—å—Ç—Ä –ø–µ—Ä–µ–¥–∞—ë—Ç—Å—è –ø–∞—Ä–∞–º–µ—Ç—Ä–æ–º
- **–ò–º–µ–Ω–æ–≤–∞–Ω–∏–µ –∫—Ä–∞—Ç–∫–æ–µ –∏ –ø—Ä–µ–¥–º–µ—Ç–Ω–æ–µ**: `fetch`, `search`, `deleteByType`, `deleteByUid`, `deleteByUids`
- **–û–±–ª–∞—Å—Ç—å –æ—Ç–≤–µ—Ç—Å—Ç–≤–µ–Ω–Ω–æ—Å—Ç–∏**: —Ç–æ–ª—å–∫–æ –æ–ø–µ—Ä–∞—Ü–∏–∏ –Ω–∞–¥ —Å–≤–æ–µ–π —Å—É—â–Ω–æ—Å—Ç—å—é. –ù–∏–∫–∞–∫–∏—Ö `uploadImages` –≤ `ThingsRepository`
- **–¢–∏–ø—ã –≤–æ–∑–≤—Ä–∞—Ç–∞ ‚Äî –º–æ–¥–µ–ª–∏** (–Ω–µ `Map<String, dynamic>`)
- **–ö–æ–Ω—Ç—Ä–∞–∫—Ç –¥–æ—Å—Ç—É–ø–∞**: **–¥–æ—Å—Ç—É–ø –∫ —Ä–µ–ø–æ–∑–∏—Ç–æ—Ä–∏—é —Ç–æ–ª—å–∫–æ —á–µ—Ä–µ–∑ –∏–Ω—Ç–µ—Ä—Ñ–µ–π—Å** (–∏–Ω–≤–µ—Ä—Å–∏—è –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–µ–π)
- **–°—Ç–∏–ª—å —Ñ–∞–π–ª–∞**: –∏–Ω—Ç–µ—Ä—Ñ–µ–π—Å —Ä–µ–ø–æ–∑–∏—Ç–æ—Ä–∏—è (–µ—Å–ª–∏ –æ–±—ä—è–≤–ª–µ–Ω –≤ —ç—Ç–æ–º –∂–µ —Ñ–∞–π–ª–µ) ‚Äî **–≤ –∫–æ–Ω—Ü–µ —Ñ–∞–π–ª–∞**; –º–µ—Ç–æ–¥ `void dispose()` (–µ—Å–ª–∏ –µ—Å—Ç—å) ‚Äî **–ø–æ—Å–ª–µ–¥–Ω–∏–º**

### üîÑ –ü–æ—Ç–æ–∫–∏ –¥–ª—è –∂–∏–≤—ã—Ö –æ–±–Ω–æ–≤–ª–µ–Ω–∏–π
–ü–æ—Ç–æ–∫–∏ –¥–ª—è –∂–∏–≤—ã—Ö –æ–±–Ω–æ–≤–ª–µ–Ω–∏–π –æ–±–µ—Å–ø–µ—á–∏–≤–∞—é—Ç—Å—è –æ—Ç–¥–µ–ª—å–Ω—ã–º–∏ —Å–µ—Ä–≤–∏—Å–∞–º–∏/–Ω–∞–±–ª—é–¥–∞—Ç–µ–ª—è–º–∏, –Ω–æ –Ω–µ –∏–Ω—Ç–µ—Ä—Ñ–µ–π—Å–æ–º —Ä–µ–ø–æ–∑–∏—Ç–æ—Ä–∏—è.

### üèóÔ∏è –ö–æ–Ω—Ç—Ä–∞–∫—Ç –¥–æ—Å—Ç—É–ø–∞
UI/BLoC –æ–±—ä—è–≤–ª—è—é—Ç –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç—å –æ—Ç `*RepositoryI`; –∫–æ–Ω–∫—Ä–µ—Ç–Ω–∞—è —Ä–µ–∞–ª–∏–∑–∞—Ü–∏—è –ø—Ä–æ–∫–∏–¥—ã–≤–∞–µ—Ç—Å—è –∫–æ–Ω—Ç–µ–π–Ω–µ—Ä–æ–º DI. –ó–∞–ø—Ä–µ—â–µ–Ω–æ –≤ UI/BLoC –∏–º–ø–æ—Ä—Ç–∏—Ä–æ–≤–∞—Ç—å –∏ —Å–æ–∑–¥–∞–≤–∞—Ç—å `new ThingsRepository(...)` –Ω–∞–ø—Ä—è–º—É—é.

---

## üìñ –ü—Ä–∏–º–µ—Ä—ã

### ‚ùå –ü–ª–æ—Ö–æ (–Ω–∞—Ä—É—à–µ–Ω–∏—è –≤ —Å–∏–≥–Ω–∞—Ç—É—Ä–∞—Ö, –∏–º–µ–Ω–∞—Ö –∏ –æ—Ç–≤–µ—Ç—Å—Ç–≤–µ–Ω–Ω–æ—Å—Ç–∏)
```dart
--8<-- "docs/examples/bad/bad-repo-state-002.dart"
```

### ‚úÖ –•–æ—Ä–æ—à–æ (–µ–¥–∏–Ω—ã–π Future-API, —Ñ–∏–ª—å—Ç—Ä—ã, –∫—Ä–∞—Ç–∫–∏–µ –∏–º–µ–Ω–∞, –∏–Ω—Ç–µ—Ä—Ñ–µ–π—Å –∏ dispose() –≤ –∫–æ–Ω—Ü–µ)
```dart
--8<-- "docs/examples/good/good-repo-state-002.dart"
```

### ‚ùå –î–æ–ø–æ–ª–Ω–∏—Ç–µ–ª—å–Ω–æ: –Ω–µ –≤–æ–∑–≤—Ä–∞—â–∞–µ–º `Map`, –∞ –∏—Å–ø–æ–ª—å–∑—É–µ–º –º–æ–¥–µ–ª–∏
```dart
--8<-- "docs/examples/bad/bad-repo-map-return-001.dart"
```

### ‚úÖ –•–æ—Ä–æ—à–æ: –≤–æ–∑–≤—Ä–∞—â–∞–µ–º –º–æ–¥–µ–ª–∏
```dart
--8<-- "docs/examples/good/good-repo-map-return-001.dart"
```

### üîí –ö–æ–Ω—Ç—Ä–∞–∫—Ç –¥–æ—Å—Ç—É–ø–∞: —Ç–æ–ª—å–∫–æ —á–µ—Ä–µ–∑ –∏–Ω—Ç–µ—Ä—Ñ–µ–π—Å

#### ‚ùå –ü–ª–æ—Ö–æ (UI/BLoC ‚Üí –∫–æ–Ω–∫—Ä–µ—Ç–Ω—ã–π –∫–ª–∞—Å—Å —Ä–µ–ø–æ–∑–∏—Ç–æ—Ä–∏—è)
```dart
--8<-- "docs/examples/bad/bad-repo-direct-access-001.dart"
```

#### ‚úÖ –•–æ—Ä–æ—à–æ (UI/BLoC ‚Üí –∏–Ω—Ç–µ—Ä—Ñ–µ–π—Å, DI –ø—Ä–æ–≤–∞–ª–∏–≤–∞–µ—Ç —Ä–µ–∞–ª–∏–∑–∞—Ü–∏—é)
```dart
--8<-- "docs/examples/good/good-repo-direct-access-001.dart"
```

---

## ü§ñ AI-–ø–æ–¥—Å–∫–∞–∑–∫–∞ –¥–ª—è –ª–∏–Ω—Ç–µ—Ä–∞

–ê–Ω–∞–ª–∏–∑–∏—Ä—É–π —Ç–æ–ª—å–∫–æ –∫–ª–∞—Å—Å—ã/–∏–Ω—Ç–µ—Ä—Ñ–µ–π—Å—ã, –∏–º—è –∫–æ—Ç–æ—Ä—ã—Ö –∑–∞–∫–∞–Ω—á–∏–≤–∞–µ—Ç—Å—è –Ω–∞ 'Repository' –∏–ª–∏ 'RepositoryI' (—Å–º. scope).

**–§–ª–∞–≥–∏—Ä—É–π:**
- init/setup/start/bootstrap/preload/configure/loadCache/warmup/open/connect ‚Äî –Ω–µ–∑–∞–≤–∏—Å–∏–º–æ –æ—Ç –ø–∞—Ä–∞–º–µ—Ç—Ä–æ–≤
- public –º–µ—Ç–æ–¥—ã —Å Stream<...>
- public void, –∫—Ä–æ–º–µ dispose
- public –º–µ—Ç–æ–¥—ã, —á—å–∏ –∏–º–µ–Ω–∞ –ù–ï –Ω–∞—á–∏–Ω–∞—é—Ç—Å—è —Å fetch|create|update|delete (–∏–≥–Ω–æ—Ä–∏—Ä—É–π private: _name, –∏ dispose)

**–ù–µ —Ñ–ª–∞–≥–∏—Ä—É–π –∫–æ–¥ –≤–Ω–µ scope.** –°–æ–æ–±—â–µ–Ω–∏–µ –∫–æ—Ä–æ—Ç–∫–æ–µ, –±–µ–∑ –Ω–µ—Ä–µ–ª–µ–≤–∞–Ω—Ç–Ω—ã—Ö —Å–æ–≤–µ—Ç–æ–≤.
